src/BuiltInTools/dotnet-watch/README.md

```diff
@@ -91,6 +91,91 @@ Example:
+...
+
+<!-- The script that follows is injected by the middleware right before the closing body tag -->
+<script src="_framework/aspnetcore-browser-refresh.js"></script>
+</body>
+```
+
+Watch uses yet another environment variable to tell the app / middleware what the URL of the server is that the injected script needs to communicate with. aspnetcore-browser-refresh.js connects to this URL and is now ready to listen to process dotnet-watch's reload commands. Prior to .NET 6.0, commands were simple strings such as 'Reload' which would cause the browser to reload or 'Wait' which would indicate that the dotnet-watch is in the process of rebuilding. Starting in 6.0, these messages also include complex JSON payloads that contain additional data. For instance, the command to update a static asset looks like:
+
+```json
+{
+  "type": "UpdateStaticFile",
+  "path": "wwwroot/css/site.css"
+}
+```
+
+The implementation for the static asset reload is heavily inspired by @RickStrah's Live Reload middleware located at https://github.com/RickStrahl/Westwind.AspnetCore.LiveReload
+
+Scoped css files (.razor.css and .cshtml.css) are a special case of static web assets. They need to be processed (currently requires running MSBuild targets) to be updated. Once updated, the mechanics of updating are identical to other static assets.
+
+#### Compiled files (.cs, .razor, .cshtml)
+
+Updating a compiled file requires two requirements -
+
+1. the compiler needs to able to produce a binary patch from the source changes
+1. The app needs to be able to apply the binary patch.
+
+The former has been available as a compiler feature to support debugger / IDE's Edit and Continue capabilities. Starting in 6.0, the runtime (.NET Core and Mono) [expose APIs](https://github.com/dotnet/runtime/blob/f15722c2c25ce945f2f1c7673ff3f4fbdb244feb/src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs#L28) to patch a running app. The combination of the two makes it possible to hot reload changes to files that are compiled as part of the application. The first bullet point is easy for files that are already compiled as part of the app e.g. .cs or .vb files. For any other file types, source generators offer a way to participate in the compilation process. This is the route taken by Razor Source Generator for .razor and .cshtml files.
+
+To enable this, dotnet-watch hosts a Roslyn workspace that contains the project and all of it's references. When a change to .cs, .razor, or .cshtml file is found, it updates the workspace and uses it's APIs to produce a patch. Not all source changes can however produce a patch. These are called rude-edits (since it rudely interrupts your workflow) and in this case, watch falls back to doing a regular build to update and restart the app. If a patch is successfuly produced, watch now has the task of sending these patches to the app. Since different app models have different constraints, this implementation is app-model specific:
```

Comment:
I think we should mention that diagnostics are used to convey why a rude-edit occurred.